Compute CoreDNS version based on Kubernetes version

kron4eg opened the pull request:
And result to deploy both, in-seed and in-user-cluster CoreDNS

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6023 

**Does this PR introduce a user-facing change?**:
```release-note
CoreDNS version now based on Kubernetes version
```

kron4eg committed:
Compute CoreDNS version based on Kubernetes version

And result to deploy both, in-seed and in-user-cluster CoreDNS

Signed-off-by: Artiom Diomin <kron82@gmail.com>

kron4eg commented:
/retest

multi-io commented:
While the cluster dns was still an addon we upgraded it to coredns 1.7.0 -- for all K8s versions. That's what e.g. kubeadm uses too, and we never had problems with that (we figured that the list at https://github.com/coredns/deployment/blob/master/kubernetes/CoreDNS-k8s_version.md may be primarily a documentation of history, not of compatibility).

In the Corefile we also removed `upstream` (which you've done in the PR now too -- but will that work with coredns 1.3.1?) and replaced `proxy . /etc/resolv.conf` with `forward . /etc/resolv.conf` because I think the proxy plugin is no longer supported in later coredns versions. Also, the `cluster.local` name shouldn't be hardcoded but set to the value of `cluster.spec.clusterNetwork.dnsDomain` if the user overrode that. -- which would require the usercluster-controller-manager to access the cluster resource, which currently isn't supported AFAIK, so for now I guess I'll be passing dnsDomain as an additional optional command line parameter to the user cluster controller manager if the user overrode it. We can also make a PR for this.

kron4eg commented:
This is what kubeadm do.

https://github.com/kubernetes/kubernetes/blob/v1.17.0/cmd/kubeadm/app/constants/constants.go#L337
```
CoreDNSVersion = "1.6.5"
```

https://github.com/kubernetes/kubernetes/blob/v1.18.0/cmd/kubeadm/app/constants/constants.go#L338
```
CoreDNSVersion = "1.6.7"
```

https://github.com/kubernetes/kubernetes/blob/v1.19.0/cmd/kubeadm/app/constants/constants.go#L332
```
CoreDNSVersion = "1.7.0"
```

https://github.com/kubernetes/kubernetes/blob/v1.20.0/cmd/kubeadm/app/constants/constants.go#L350
```
CoreDNSVersion = "1.7.0"
```

https://github.com/kubernetes/kubernetes/blob/v1.21.0-alpha.2/cmd/kubeadm/app/constants/constants.go#L350
```
CoreDNSVersion = "v1.8.0"
```

kron4eg commented:
> but will that work with coredns 1.3.1

Yes, because there will be no coredns 1.3.1 :) Oldest version will be 1.6.5

kron4eg commented:
/retest

kron4eg commented:
> and replaced `proxy . /etc/resolv.conf` with `forward . /etc/resolv.conf` because I think the proxy plugin is no longer supported in later coredns versions.

Yeah, replaced that too.

youssefazrak reviewed (approved):
/lgtm

youssefazrak commented:
@kron4eg @multi-io  FYI `upstream` is the default behavior since CoreDNS 1.4 and later. Before those versions, this is indeed needed.
Regarding the `proxy`, Agree here, deprecated this 1.5.0 (https://coredns.io/2019/04/06/coredns-1.5.0-release/)

kron4eg commented:
/retest

kron4eg commented:
/retest
